neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
 
         self._rank = rank
+        self._num_sessions = num_sessions
         self._all_encoders = all_encoders
-        self._bs_outputs = bs_outputs
-        self._vocabulary = vocabulary
+        self._decoder = decoder
         self._postprocess = postprocess
 
-        self.result = None  # type: ExecutionResult
+        self.step = 0
+        self.scores = np.empty([0, decoder.beam_size], dtype=float)
+        self.parent_ids = np.empty([0, decoder.beam_size], dtype=int)
+        self.token_ids = np.empty([0, decoder.beam_size], dtype=int)
 
-    def next_to_execute(self) -> NextExecute:
-        return self._all_encoders, {"bs_outputs": self._bs_outputs}, {}
+        self._next_feed = [{} for _ in range(self._num_sessions)] \
+            # type: List[FeedDict]
 
-    def collect_results(self, results: List[Dict]) -> None:
-        if len(results) > 1:
-            raise ValueError("Beam search runner does not support ensembling.")
+        # If we are ensembling, we run only one step at a time
+        # Note, that in the first iteration, we only want to generate
+        # logprobs for ensembling, therefore we set max_steps = 0
+        if self._num_sessions > 1:
+            for fd in self._next_feed:
+                fd.update({self._decoder.max_steps: 0})
 
-        evaluated_bs = results[0]["bs_outputs"]
-        max_time = evaluated_bs.scores.shape[0]
+        self.result = None  # type: ExecutionResult
 
-        # pick the end of the hypothesis based on its rank
-        hyp_index = np.argpartition(
-            -evaluated_bs.scores[-1], self._rank - 1)[self._rank - 1]
-        bs_score = evaluated_bs.scores[-1][hyp_index]
+    def next_to_execute(self) -> NextExecute:
+        return (self._all_encoders,
+                {"bs_outputs": self._decoder.outputs},
+                self._next_feed)
 
-        # now backtrack
-        output_tokens = []  # type: List[str]
+    # pylint: disable=too-many-locals
+    def collect_results(self, results: List[Dict]) -> None:
+        # Recompute logits
+        # Only necessary when ensembling models
+        prev_logprobs = [res["bs_outputs"].last_search_state.prev_logprobs
+                         for res in results]
+
+        # Arithmetic mean
+        ens_logprobs = (scipy.misc.logsumexp(prev_logprobs, 0)
+                        - np.log(self._num_sessions))
+
+        # The last score is computed
+        bs_outputs = results[0]["bs_outputs"]
+        step = bs_outputs.last_dec_loop_state.step - 1
+
+        self.step += step
+        self.scores = np.append(
+            self.scores,
+            bs_outputs.last_search_step_output.scores[0:step],
+            axis=0)
+
+        self.parent_ids = np.append(
+            self.parent_ids,
+            bs_outputs.last_search_step_output.parent_ids[0:step],
+            axis=0)
+
+        self.token_ids = np.append(
+            self.token_ids,
+            bs_outputs.last_search_step_output.token_ids[0:step],
```

Comment:
proč to jde pokaždý od nuly? nekumulujou se pak v těch maticích ty stejný věci?

Reply:
Viz predchozi komentar. U normalniho behu se tohle vola jednou a parametry jsou 0:max_seq_len
U ensemblu se to vola v kazdem kroku a vzhledem k tomu, ze to poustime po jednotlivych krocich, tak je to jen 0:1